### ISSUE TITLE
cmd/compile: simplify multiplication and division of Abs() 

### ISSUE THREAD

**Issue by Jacalz** [golang/go#50126]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.16.8 linux/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes. See https://go.godbolt.org/z/1Gq4bPPnd for assembly output.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/jacob/.cache/go-build"
GOENV="/home/jacob/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/jacob/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/jacob/go"
GOPRIVATE=""
GOPROXY="direct"
GOROOT="/usr/lib/golang"
GOSUMDB="off"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/golang/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.8"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3228184624=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

I noticed that the compile doesn't optimize calls to `math.Abs()` or `cmplx.Abs()` to strip out unnecessary calls to those functions when multiplying them or dividing them. According to the absolute value rules, `|A * B| = |A| * |B|` and `|A / B| = |A| / |B|`, meaning that the compiler should be able to work the right two calls to either `math.Abs()` or `cmplx.Abs()` into a single call by moving the multiplication or division inside the call instead.

This means that the first block of assembly can be simplified to the one after.

```asm
        MOVQ    X0, AX
        BTRQ    $63, AX
        MOVQ    AX, X0
        MOVQ    X1, AX
        BTRQ    $63, AX
        MOVQ    AX, X1
        MULSD   X1, X0
```

```asm
        MULSD   X1, X0
        MOVQ    X0, AX
        BTRQ    $63, AX
        MOVQ    AX, X0
```

The above assembly was taken from the multiplication examples below:

```go
func AbsMult(x, y float64) float64 {
    return math.Abs(x) * math.Abs(y)
}

func AbsMultSimple(x, y float64) float64 {
    return math.Abs(x * y)
}

func AbsDiv(x, y float64) float64 {
    return math.Abs(x) / math.Abs(y)
}

func AbsDivSimple(x, y float64) float64 {
    return math.Abs(x / y)
}
```

This should also apply to complex numbers. It is important to not that it is only for multiplication and division, not addition and subtraction.

### What did you expect to see?
I expected the compiler to strip out the duplicated calls to `math.Abs()`. Thus getting the following assembly for the `AbsMult` function below.

```asm
        MULSD   X1, X0
        MOVQ    X0, AX
        BTRQ    $63, AX
        MOVQ    AX, X0
```
```go
func AbsMult(x, y float64) float64 {
    return math.Abs(x) * math.Abs(y)
}

func AbsMultSimple(x, y float64) float64 {
    return math.Abs(x * y)
}

func AbsDiv(x, y float64) float64 {
    return math.Abs(x) / math.Abs(y)
}

func AbsDivSimple(x, y float64) float64 {
    return math.Abs(x / y)
}
```

### What did you see instead?

The compiler did not strip it out:

```asm
        MOVQ    X0, AX
        BTRQ    $63, AX
        MOVQ    AX, X0
        MOVQ    X1, AX
        BTRQ    $63, AX
        MOVQ    AX, X1
        MULSD   X1, X0
```

For more info, see: https://github.com/dominikh/go-tools/issues/1135

**Comment by ALTree** [comment]
The optimization may be correct, but is it worthwhile to hardcode it in the compiler? You can just write `math.Abs(x * y)`, after all.

I can understand special-casing arithmetic simplifications for built-in operators (like `+` or `*`), but I'm not sure about `math` functions.

Even gcc doesn't do this on `-O2`:

```
#include <math.h>
double f1(double i, double j) { return fabs(i)*fabs(j); }
double f2(double i, double j) { return fabs(i*j); }
```

```
f1:
        movq    xmm2, QWORD PTR .LC0[rip]
        andpd   xmm0, xmm2
        andpd   xmm1, xmm2
        mulsd   xmm0, xmm1
        ret
f2:
        mulsd   xmm0, xmm1
        andpd   xmm0, XMMWORD PTR .LC0[rip]
        ret
```

Clang, on the other hand, simplifies it, but still...

**Comment by Jacalz** [comment]
Indeed. That is a good question that also was brought up in the issue that I opened for staticcheck. I figured that it was better to open an issue here and let you all decide on that :)
